contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,384 @@
+    mapping (uint256 => ServiceInfo) public mapServiceInfo;
+    // Set of currently staking serviceIds
+    uint256[] public setServiceIds;
+
+    /// @dev ServiceStakingBase constructor.
+    /// @param _rewardsPerSecond Staking rewards per second (in single digits).
+    /// @param _minStakingDeposit Minimum staking deposit for a service to be eligible to stake.
+    /// @param _livenessRatio Staking ratio: number of seconds per nonce (in 18 digits).
+    /// @param _serviceRegistry ServiceRegistry contract address.
+    constructor(uint256 _rewardsPerSecond, uint256 _minStakingDeposit, uint256 _livenessRatio, address _serviceRegistry) {
+        // Initial checks
+        if (_rewardsPerSecond == 0 || _minStakingDeposit == 0 || _livenessRatio == 0) {
+            revert ZeroValue();
+        }
+        if (_serviceRegistry == address(0)) {
+            revert ZeroAddress();
+        }
+
+        rewardsPerSecond = _rewardsPerSecond;
+        minStakingDeposit = _minStakingDeposit;
+        livenessRatio = _livenessRatio;
+        serviceRegistry = _serviceRegistry;
+    }
+
+    /// @dev Checks token / ETH staking deposit.
+    /// @param stakingDeposit Staking deposit.
+    function _checkTokenStakingDeposit(uint256, uint256 stakingDeposit) internal view virtual {
+        // The staking deposit derived from a security deposit value must be greater or equal to the minimum defined one
+        if (stakingDeposit < minStakingDeposit) {
+            revert LowerThan(stakingDeposit, minStakingDeposit);
+        }
+    }
+
+    /// @dev Withdraws the reward amount to a service owner.
+    /// @param to Address to.
+    /// @param amount Amount to withdraw.
+    function _withdraw(address to, uint256 amount) internal virtual {}
+
+    /// @dev Stakes the service.
+    /// @param serviceId Service Id.
+    function stake(uint256 serviceId) external {
+        // Check the service conditions for staking
+        (uint96 stakingDeposit, address multisig, , , , , uint8 state) = IService(serviceRegistry).mapServices(serviceId);
+        // The service must be deployed
+        if (state != 4) {
+            revert WrongServiceState(state, serviceId);
+        }
+        // Check service staking deposit and token, if applicable
+        _checkTokenStakingDeposit(serviceId, stakingDeposit);
+
+        // Transfer the service for staking
+        IService(serviceRegistry).transferFrom(msg.sender, address(this), serviceId);
+
+        // ServiceInfo struct will be an empty one since otherwise the transferFrom above would fail
+        ServiceInfo storage sInfo = mapServiceInfo[serviceId];
+        sInfo.multisig = multisig;
+        sInfo.owner = msg.sender;
+        sInfo.nonce = IMultisig(multisig).nonce();
+        sInfo.tsStart = block.timestamp;
+
+        // Add the service Id to the set of staked services
+        setServiceIds.push(serviceId);
+
+        emit ServiceStaked(serviceId, msg.sender);
```

Comment:
Can we also emit the service multisig address and the nonce at the time of the stake?

Reply:
why?